sodetlib/quality_control.py

```diff
@@ -0,0 +1,142 @@
+import sodetlib as sdl
+import numpy as np
+import time
+from pysmurf.client.base.smurf_control import SmurfControl
+
+def check_packet_loss(Ss, cfgs, dur=10, fr_khz=4, nchans=2000, slots=None):
+    """
+    Takes a short G3 Stream on multiple slots simultaneously and checks for
+    dropped samples. This function is strange since it requires simultaneous
+    streaming on multiple slots to properly test, so it doesn't follow the
+    standard sodetlib function / data format.
+
+    Args
+    -----
+    Ss : dict[SmurfController]1
+        Dict of pysmurf instances where the key is the slot-number
+    cfgs : dict[DetConfig]
+        Dict of DetConfigs where the key is the slot number
+    dur : float
+        Duration of data stream (sec)
+    fr_khz : float
+        Frequency of FR rate (khz)
+    nchans : int
+        Number of channels to stream
+    slots : list
+        Which slots to stream data on. If None, will stream on all slots in the
+        Ss object.
+
+    Returns
+    --------
+    ams : dict[AxisManagers]
+        Dict of axis managers indexed by slot-number
+    res : dict
+        Dict where the key is the slot number, and the values are dicts
+        containing frame counters, number of dropped frames, etc.
+    """
+    if slots is None:
+        slots = Ss.keys()
+
+    for s in slots:
+        S = Ss[s]
+        S.flux_ramp_setup(fr_khz, 0.4, band=0)
```

Comment:
this doesn't require tones to be setup, if you try streaming without tones setup it will just stream garbage data which is fine for this check. The flux_ramp_setup is just to set the data rate.